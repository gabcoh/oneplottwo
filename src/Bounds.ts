/*
 * TODO improve this header with documentation
 */
// Not really sure how I'm going to implement different types of graphing, but
// its probably not a bad idea to specify rectangular bounds here now.
export interface RectangularBounds {
  minX: number;
  maxX: number;

  minY: number;
  maxY: number;

  minZ: number;
  maxZ: number;
}
