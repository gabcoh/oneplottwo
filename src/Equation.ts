
// Really just cartesian but generalize later (when you even figure out how to plot other curves)
// TODO Make this not so naive. Actually parse equation into something real (DONT JUST EVAL IDIOT)
export class Equation {
  rawEquation: string;

  constructor(rawEquation: string) {
    this.rawEquation = rawEquation;
  }
  evaluate(x: number, y: number) {
    return eval(this.rawEquation);
  }
  derivate(bx: number, by: number, delta: number): [number, number] {
    let x = bx;
    let y = by;
    const fxy = eval(this.rawEquation);
    x = x + delta;
    const dfx = eval(this.rawEquation) - fxy;
    x = bx;
    y = y + delta;
    const dfy = eval(this.rawEquation) - fxy;
    return [dfx / delta, dfy / delta];
  }
}
