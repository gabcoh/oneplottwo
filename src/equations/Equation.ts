/*
 * TODO improve this header
 */
export interface RectangularEquation {
  rawEquation: string;

  evaluate(x: number): number;
  derivate(bx: number, by: number, delta: number): [number, number];
}
