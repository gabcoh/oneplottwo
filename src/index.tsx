import * as React from 'react';
import * as ReactDOM from 'react-dom';

import '../public/index.html';
import { Grapher } from './Grapher';

import { GraphController } from './ui/GraphController';

const grapher = new Grapher(document.getElementById('graph') as HTMLElement);
grapher.animate();

function onResizeDone() {
  let timeout: number;
  return function () {
    clearTimeout(timeout);
    // MUST USE window. to avoid confusion with node settimeout
    timeout = window.setTimeout(() => grapher.updateSize(), 250);
  };
}
window.addEventListener('resize', onResizeDone());

ReactDOM.render(
  <GraphController
    grapher={grapher}
  />,
  document.getElementById('ui') as HTMLElement,
);

// TODO this is a little hacky. maybe fix it
class DragHandlerClass {
  down: boolean;
  ui: HTMLElement;
  graph: HTMLElement;
  bar: HTMLElement;
  grapher: Grapher;
  constructor(grapher: Grapher) {
    this.grapher = grapher;
    this.down = false;
    this.onMouseMove = this.onMouseMove.bind(this);
    this.onMouseUp = this.onMouseUp.bind(this);
    this.onMouseDown = this.onMouseDown.bind(this);
    this.graph = document.getElementById('graph') as HTMLElement;
    this.ui = document.getElementById('ui') as HTMLElement;
    this.bar = document.getElementById('bar') as HTMLElement;
  }
  onMouseMove(e: Event) {
    if (!this.down) {
      return;
    }
    const fraction =
      ((e as MouseEvent).clientX - this.bar.clientWidth / 2) / document.body.clientWidth;
    this.graph.style.width = `${fraction * 100}%`;
    this.ui.style.width = `${99 - fraction * 100}%`;
  }
  onMouseDown(e: Event) {
    if (e.target !== this.bar) {
      return;
    }
    e.preventDefault();
    this.down = true;
  }
  onMouseUp(e: Event) {
    if (this.down === true) {
      console.log('up');
      grapher.updateSize();
      this.down = false;
    }
  }
}
const dh = new DragHandlerClass(grapher);
document.addEventListener('mousedown', dh.onMouseDown);
document.addEventListener('mouseup', dh.onMouseUp);
document.addEventListener('mousemove', dh.onMouseMove);
