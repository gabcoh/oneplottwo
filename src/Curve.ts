import * as THREE from 'three';

import { RectangularBounds } from './Bounds';
import { Equation } from './Equation';
// This is really just for functions of z=f(x,y) but I'll deal with that later
// Also, maybe call this Curve and the actual function Function
export class Curve {
  // Number of points in each direction for mesh
  points: number;
  // Just a string that gets evalled for now, should soon become a more complicated parsed object
  equation: Equation;

  // TODO IMPLEMENT BOUNDS OBJECT
  bounds: RectangularBounds = {
    minX: -1,
    maxX: 1,
    minY: -1,
    maxY: 1,
    minZ: -1,
    maxZ: 1,
  };

  geometry: THREE.BufferGeometry;
  material: THREE.Material;
  mesh: THREE.Object3D;

  constructor() {
    this.equation = new Equation('x*y');
    // Set to a reasonable default
    this.points = 10;

    this.geometry = new THREE.BufferGeometry();
    this.updateBuffers();
    // Must explicitly set to infinity for whatever reason
    // This caused hours of headache...
    this.geometry.setDrawRange(0, Infinity);

    this.material = new THREE.MeshPhongMaterial({
      side: THREE.DoubleSide,
      specular: 0xffffff,
      shininess: 1,
      wireframe: false,
      vertexColors: THREE.VertexColors,
    });
    this.mesh = new THREE.Mesh(this.geometry, this.material);
  }

  // only need to update indicies when number of points changes
  updateBuffers() {
    const [vertices, colors, normals] = this.generatePointsColorsAndNormals();
    this.geometry.addAttribute('position', new THREE.BufferAttribute(vertices, 3));
    this.geometry.addAttribute('normal', new THREE.BufferAttribute(normals, 3));
    this.geometry.addAttribute('color', new THREE.BufferAttribute(colors, 3));
    const indicies = this.generateIndicies();
    this.geometry.setIndex(new THREE.BufferAttribute(indicies, 1));
  }
  generatePointsColorsAndNormals(): [Float32Array, Float32Array, Float32Array] {
    const vertices = new Float32Array(this.points * this.points * 3);
    const normals = new Float32Array(this.points * this.points * 3);
    const colors = new Float32Array(this.points * this.points * 3);
    const xRange = this.bounds.maxX - this.bounds.minX;
    const yRange = this.bounds.maxY - this.bounds.minY;
    for (let i = 0; i < this.points; i += 1) {
      for (let j = 0; j < this.points; j += 1) {
        const base = i * this.points * 3 + j * 3;
        const x = (j / (this.points - 1)) * xRange + this.bounds.minX;
        const y = (i / (this.points - 1)) * yRange + this.bounds.minY;

        vertices[base + 0] = x;
        vertices[base + 1] = y;
        vertices[base + 2] = this.equation.evaluate(x, y);

        const [dx, dy] = this.equation.derivate(x, y, xRange / 1000);
        const vx = new THREE.Vector3(1, 0, dx);
        const vy = new THREE.Vector3(0, 1, dy);
        vx.cross(vy);
        normals[base + 0] = vx.x;
        normals[base + 1] = vx.y;
        normals[base + 2] = vx.z;

        colors[base + 0] = (x - this.bounds.minX) * 255 / xRange;
        colors[base + 1] = (y - this.bounds.minY) * 255 / yRange;
        colors[base + 2] = 0;
      }
    }
    return [vertices, colors, normals];
  }
  generateIndicies(): Uint32Array {
    const indicies = new Uint32Array((this.points - 1) * (this.points - 1) * 2 * 3);
    for (let i = 0; i < this.points - 1; i += 1) {
      for (let j = 0; j < this.points - 1; j += 1) {
        const base = i * (this.points - 1) * 6 + (j * 6);
        const a = i * this.points + j + 1;
        const b = i * this.points + j;
        const c = (i + 1) * this.points + j;
        const d = (i + 1) * this.points + j + 1;

        indicies[base + 0] = d;
        indicies[base + 1] = b;
        indicies[base + 2] = a;

        indicies[base + 3] = d;
        indicies[base + 4] = c;
        indicies[base + 5] = b;
      }
    }
    return indicies;
  }
  dispose() {
    this.geometry.dispose();
    this.material.dispose();
  }
  updateEquation(rawEquation: string) : (Error | null) {
    try {
      this.equation = new Equation(rawEquation);
    } catch (e) {
      return e;
    }
    this.updateBuffers();
    return null;
  }
}
