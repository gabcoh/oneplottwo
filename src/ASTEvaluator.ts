/*
 * TODO implement it
 *
 * this file implements everything neceseery to evalue an ast.
 */
import { isAnASTFuncNode, ASTNode } from './EquationParser';

export function evaluateAST(ast: ASTNode, vars: Map<string, number>): number {
  if (isAnASTFuncNode(ast)) {
    const params: (Error | number)[] = ast.operands.map(a => evaluateAST(a, vars));
    for (let i = 0; i < params.length; i = i + 1) {
      if (typeof params[i] !== 'number') {
        throw  new Error(`${params[i]} is not a number`);
      }
    }
    const valOrError = ast.operator.func(...params as number[]);
    if (valOrError instanceof Error) {
      throw valOrError;
    }
    return valOrError as number;
  }
  if (typeof ast === 'string') {
    if (vars.has(ast)) {
      return vars.get(ast) as number;
    }
    throw new Error(`variable ${ast} is undefined`);
  }
  if (typeof ast === 'number') {
    return ast;
  }
  throw new Error('unknown type of astnode');
}
