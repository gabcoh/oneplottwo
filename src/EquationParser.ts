/*
 * TODO Improve error messages
 * TODO throw errors in lexer if appropriate
 * TODO consider returning errors so that typescript knows there type
 * TODO maybe let each curve/equation class implement parse equation
 * TODO parse equation should return an equation. Not an ast
 *
 * This file implements dikstra's shunting yard algorithm to parse arithmetic expressions.
 * Raw equations are first passed to the lexer which tokenizes them and labels each token,
 * with it's type (eg. variable/function, constant, operator). The tokens are then passed
 * to the shunting yard which parses the tokens into an AST and throws some errors in the
 * case of malformed equations.
 *
 * So far, equations can consist of functions, parenthized expressions, and binary operators.
 * Prefix operators should work, but none are implemented and for that reasons there also
 * are no tests for them.
 * Names of functions and variables can not contain numbers!
 *
 */
import { EquationTypes } from './equations/EquationTypes';
import { ExplicitRectangularEquation } from './equations/RectangularEquations';
import { Equation } from './equations/Equation';

export enum Associativity {
  LEFT,
  RIGHT,
}

export interface Operator {
  name: string;
  precedence: number;
  // have the option to return errro but I am not using that at the moment
  func: (...params: number[]) => (Error | number);
  associativity?: Associativity;
  arity: number;
  prefix?: boolean;
}
const PAREN_OPERATOR: Operator = {
  name: 'PAREN',
  precedence: -1,
  arity: -1,
  func(...params: number[]) {
    return new Error('You should never call = as function');
  },
};
const functions: Map<string, Operator> = new Map([
  ['sum', {
    name: 'EXP',
    precedence: 4,
    arity: 2,
    func(...params: number[]) {
      return params.reduce((a, b) => a + b);
    },
  }],
]);
const operators: Map<string, Operator> = new Map([
  ['^', {
    name: 'EXP',
    precedence: 4,
    arity: 2,
    associativity: Associativity.RIGHT,
    func(...params: number[]) {
      return Math.pow(params[0], params[1]);
    },
  }],
  ['*', {
    name: 'MULT',
    precedence: 3,
    arity: 2,
    associativity: Associativity.RIGHT,
    func(...params: number[]) {
      return params[0] * params[1];
    },
  }],
  ['/', {
    name: 'DIV',
    precedence: 3,
    arity: 2,
    associativity: Associativity.RIGHT,
    func(...params: number[]) {
      return params[0] / params[1];
    },
  }],
  ['+', {
    name: 'ADD',
    precedence: 2,
    arity: 2,
    associativity: Associativity.RIGHT,
    func(...params: number[]) {
      return params[0] + params[1];
    },
  }],
  ['-', {
    name: 'SUB',
    precedence: 2,
    arity: 2,
    associativity: Associativity.RIGHT,
    func(...params: number[]) {
      return params[0] - params[1];
    },
  }],
  ['=', {
    name: 'EQ',
    precedence: 1,
    arity: 2,
    associativity: Associativity.RIGHT,
    func(...params: number[]) {
      return new Error('You should never call = as function');
    },
  }],
]);

export enum TokenTypes {
  OPERATOR,
  NAME,
  CONSTANT,
  COMMA,
  PAREN,
}

export interface Token {
  kind: TokenTypes;
  value: string;
  column: number;
}
function lexer(raw: string): Token[] {
  let currentToken: string = '';
  let currentTokenStart: number = 0;
  const tokens: Token[] = [];
  for (let i = 0; i < raw.length; i = i + 1) {
    if (/[() ,]/.test(raw[i]) || operators.has(raw[i])) {
      if (/[0-9]+/.test(currentToken)) {
        tokens.push({
          kind: TokenTypes.CONSTANT,
          value: currentToken,
          column: currentTokenStart,
        });
      } else if (currentToken !== '') {
        tokens.push({
          kind: TokenTypes.NAME,
          value: currentToken,
          column: currentTokenStart,
        });
      }
      if (operators.has(raw[i])) {
        tokens.push({
          kind: TokenTypes.OPERATOR,
          value: raw[i],
          column: i,
        });
      } else if (/[,()]/.test(raw[i])) {
        tokens.push({
          kind: raw[i] === ',' ? TokenTypes.COMMA : TokenTypes.PAREN,
          value: raw[i],
          column: i,
        });
      }
      currentToken = '';
      currentTokenStart = i + 1;
    } else {
      currentToken += raw[i];
    }
  }
  if (/[0-9]+/.test(currentToken)) {
    tokens.push({
      kind: TokenTypes.CONSTANT,
      value: currentToken,
      column: currentTokenStart,
    });
  } else if (currentToken !== '') {
    tokens.push({
      kind: TokenTypes.NAME,
      value: currentToken,
      column: currentTokenStart,
    });
  }
  return tokens;
}

export interface ASTFuncNode {
  operator: Operator;
  operands: ASTNode[];
}
export type ASTNode = number | string | ASTFuncNode;

function applyNextOperator(operatorStack: Operator[], operandStack: ASTNode[]) {
  if (operatorStack.length === 0) {
    throw new Error('inssuficient operators');
  }
  const topOp: Operator = operatorStack.pop() as Operator;
  if (topOp.arity > operandStack.length) {
    throw new Error('insuffiecnt operands');
  }
  const ops = [];
  for (let i = 0; i < topOp.arity; i = i + 1) {
    ops.unshift(operandStack.pop() as ASTNode);
  }
  operandStack.push({
    operator: topOp,
    operands: ops,
  });
}
// Just so i don't forget a function is a name followed by a paren,
// a var is a name not followed by a paren
function shuntingYard(tokens: Token[]): ASTNode {
  const operatorStack: Operator[] = [];
  const operandStack: ASTNode[] = [];
  const functionParens = new Set();
  const functionOperandStack: {
    operands: ASTNode[],
  }[] = [];
  let previousToken: (Token | undefined) = undefined;
  for (let i = 0; i < tokens.length; i = i + 1) {
    switch (tokens[i].kind) {
      case (TokenTypes.CONSTANT):
        operandStack.push(Number(tokens[i].value));
        break;
      case(TokenTypes.NAME):
        operandStack.push(tokens[i].value);
        break;
      case(TokenTypes.OPERATOR):
        // Lexer already checked that token exists in operators map
        const currentOp = operators.get(tokens[i].value) as Operator;
        if (previousToken === undefined ||
          (previousToken as Token).kind === TokenTypes.OPERATOR &&
          currentOp.arity === 1 && !(currentOp.prefix as boolean)) {
          throw new Error('incorrect use of unary operator');
        }
        while (true) {
          if (operatorStack.length !== 0 &&
            !(currentOp.arity === 1 && operatorStack[operatorStack.length - 1].arity > 1) &&
            operatorStack[operatorStack.length - 1].precedence >= currentOp.precedence) {
            applyNextOperator(operatorStack, operandStack);
          } else {
            break;
          }
        }
        operatorStack.push(currentOp);
        break;
      case(TokenTypes.PAREN):
        if (tokens[i].value === '(') {
          operatorStack.push(PAREN_OPERATOR);
          if ((previousToken as Token).kind === TokenTypes.NAME) {
            functionOperandStack.push({
              operands: [],
            });
            functionParens.add(operatorStack.length);
          }
        } else {
          while (operatorStack[operatorStack.length - 1] !== PAREN_OPERATOR) {
            applyNextOperator(operatorStack, operandStack);
          }
          operatorStack.pop();
          if (operandStack.length === 0) {
            throw new Error('function error!!!');
          }
          // +1 because poped paren off already
          if (functionParens.delete(operatorStack.length + 1)) {
            functionOperandStack[functionOperandStack.length - 1].operands.push(
              operandStack.pop() as ASTNode);
            const funcOperands = functionOperandStack[functionOperandStack.length - 1].operands;
            const fun = operandStack.pop() as string;
            // I can't remember if this was already checked
            if (!functions.has(fun)) {
              throw new Error(`function ${fun} is not defined`);
            }
            operandStack.push({
              operator: functions.get(fun) as Operator,
              operands: funcOperands,
            });
            functionOperandStack.pop();
          }
        }
        break;
      case(TokenTypes.COMMA):
        while (operatorStack[operatorStack.length - 1] !== PAREN_OPERATOR) {
          applyNextOperator(operatorStack, operandStack);
        }
        if (operandStack.length === 0) {
          throw new Error('empty function argument');
        }
        functionOperandStack[functionOperandStack.length - 1].operands.push(
          operandStack.pop() as ASTNode);
        break;
    }
    previousToken = tokens[i];
  }
  // This is repeated from above
  while (operatorStack.length !== 0) {
    if (operatorStack[operatorStack.length - 1] === PAREN_OPERATOR) {
      throw new Error('unmatched left paren');
    }
    applyNextOperator(operatorStack, operandStack);
  }
  if (operatorStack.length !== 0 || operandStack.length !== 1) {
    throw new Error('equation error');
  }

  return operandStack[0];
}

export function isAnASTFuncNode(obj: any): obj is ASTFuncNode {
  return typeof obj === 'object' && 'operator' in obj && 'operands' in obj;
}
// Remember if mor etypes are added to ASTNode to do propper checking
function countVariables(ast: ASTNode): string[] {
  if (typeof ast === 'string') {
    return [ast];
  }
  if (isAnASTFuncNode(ast)) {
    let vars: string[] = [];
    for (let i = 0; i < ast.operands.length; i = i + 1) {
      vars = vars.concat(countVariables(ast.operands[i]));
    }
    return vars;
  }
  return [];
}
export function parseEquation(rawEquation: string): (Equation | Error) {
  try {
    const ast = shuntingYard(lexer(rawEquation));
    if (!(isAnASTFuncNode(ast)) ||
      (ast as ASTFuncNode).operator !== (operators.get('=') as Operator)) {
      return new Error('Equation does not have two expressions set equal');
    }
    // If there is not a side with a single variable (out of x, y, z) on it then
    // treat it as implicit
    const leftVars = new Set(countVariables(ast.operands[0]));
    const rightVars = new Set(countVariables(ast.operands[1]));
    const allVars = new Set([...leftVars, ...rightVars]);

    // Figure out where independent var is and what it is
    if (leftVars.size === 1 && (leftVars.has('x') || leftVars.has('X'))) {
      return new ExplicitRectangularEquation(rawEquation, 'x', ast.operands[1]);
    }
    if (leftVars.size === 1 && (leftVars.has('z') || leftVars.has('Z'))) {
      return new ExplicitRectangularEquation(rawEquation, 'z', ast.operands[1]);
    }
    if (leftVars.size === 1 && (leftVars.has('y') || leftVars.has('Y'))) {
      return new ExplicitRectangularEquation(rawEquation, 'y', ast.operands[1]);
    }
    if (rightVars.size === 1 && (rightVars.has('x') || rightVars.has('X'))) {
      return new ExplicitRectangularEquation(rawEquation, 'x', ast.operands[0]);
    }
    if (rightVars.size === 1 && (rightVars.has('z') || rightVars.has('Z'))) {
      return new ExplicitRectangularEquation(rawEquation, 'z', ast.operands[0]);
    }
    if (rightVars.size === 1 && (rightVars.has('y') || rightVars.has('Y'))) {
      return new ExplicitRectangularEquation(rawEquation, 'y', ast.operands[0]);
    }
    return new Error(`cannot parse ${rawEquation} as Rectangular`);
  } catch (e) {
    if (e instanceof Error) {
      return e;
    }
    throw e;
  }
}
const __PRIVATE__ = {
  lexer,
  operators,
  functions,
  shuntingYard,
  countVariables,
};
export {
  __PRIVATE__,
};
