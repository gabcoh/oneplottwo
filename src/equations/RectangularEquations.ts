/*
 * TODO improve this header
 * TODO fix error handling in eval
 */

import { ASTNode } from '../EquationParser';
import { evaluateAST } from '../ASTEvaluator';

export class ExplicitRectangularEquation {
  kind: string[] = ['explicit', 'rectangular'];

  rawEquation: string;
  independentVariable: string;

  ast: ASTNode;

  constructor(rawEquation: string, ind: string, ast: ASTNode) {
    this.rawEquation = rawEquation;
    this.independentVariable = ind;
    this.ast = ast;
  }
  evaluate(a: number, b: number) {
    let vars = new Map<string, number>([['y', a], ['z', b]]);
    if (this.independentVariable === 'z') {
      vars = new Map<string, number>([['x', a], ['y', b]]);
    } else if (this.independentVariable === 'y') {
      vars = new Map<string, number>([['x', a], ['z', b]]);
    }
    return evaluateAST(this.ast, vars);
  }
  derivate(x: number, y: number, delta: number): [number, number] {
    const vars = new Map<string, number>([['x', x], ['y', y]]);
    const fxy = evaluateAST(this.ast, vars);
    vars.set('x', x + delta);
    const dfx = evaluateAST(this.ast, vars) - fxy;
    vars.set('x', x);
    vars.set('y', y + delta);
    const dfy = evaluateAST(this.ast, vars) - fxy;
    return [dfx / delta, dfy / delta];
  }
}
