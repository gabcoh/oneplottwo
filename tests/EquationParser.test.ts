
import { real, MersenneTwister19937 } from 'random-js';

import * as parser from '../src/EquationParser';
import { ExplicitRectangularEquation } from '../src/equations/RectangularEquations';

const lexer = parser.__PRIVATE__.lexer;
const shuntingYard = parser.__PRIVATE__.shuntingYard;
const countVariables = parser.__PRIVATE__.countVariables;
const ops = parser.__PRIVATE__.operators;
const functions = parser.__PRIVATE__.functions;

// Make token (return object to ignore column because that would be annoying)
function mkt(t: parser.TokenTypes, s: string): Object {
  return {
    kind: t,
    value: s,
  };
}
// Make constant
function mkc(s: string): Object {
  return mkt(parser.TokenTypes.CONSTANT, s);
}
// Make operator 
function mko(s: string): Object {
  return mkt(parser.TokenTypes.OPERATOR, s);
}
const lp = mkt(parser.TokenTypes.PAREN, '(');
const rp = mkt(parser.TokenTypes.PAREN, ')');
const c =  mkt(parser.TokenTypes.COMMA, ',');
describe('lexer', () => {
  test('correctly lexes', () => {
    let expected = [mkc('1'), mko('+'), mkc('2')];
    expect(lexer('1+2')).toMatchObject(expected);
    expect(lexer(' 1 +2')).toMatchObject(expected);
    expect(lexer(' 1 + 2 ')).toMatchObject(expected);

    expected = [mkc('1'), mko('+'), lp, mkc('2'), mko('*'), mkc('3'), rp];
    expect(lexer('1+(2*3)')).toMatchObject(expected);
    expect(lexer('   1+  (  2* 3 )')).toMatchObject(expected);
    expected = [mkc('1'), mko('+'), lp, mkc('2'), mko('*'), mkt(parser.TokenTypes.NAME, 'fake'), 
      lp, mkc('3'), c, mkc('1'), mko('+'), mkc('2'), rp, rp];
    expect(lexer('1+(2*fake(3, 1+2))')).toMatchObject(expected);
    expect(lexer('1 +(2 *fake (  3,  1+ 2)   )')).toMatchObject(expected);
  });
});
describe('shunting yard', () => {
  test('correctly shunts simple expressions', () => {
    let expected: parser.ASTNode = {
      operator: ops.get('+') as parser.Operator,
      operands: [1, 2],
    };
    expect(shuntingYard(lexer('1+2'))).toEqual(expected);
    expect(() => {
      shuntingYard(lexer('1+ 2 +'))
    }).toThrow();
  });
  test('correctly shunts complex expressions', () => {
    // TODO better testing
    let expected: parser.ASTNode = {
      operator: ops.get('=') as parser.Operator,
      operands: [7, {
        operator: ops.get('+') as parser.Operator,
        operands: [1, {
          operator: ops.get('*') as parser.Operator,
          operands: [2, {
            operator: ops.get('^') as parser.Operator,
            operands: [3, 2],
          },]
        },]
      },],
    };
    expect(shuntingYard(lexer('7 = 1+2 * 3^2'))).toEqual(expected);
  });
  test('correctly shunts parenthised expressions', () => {
    let expected: parser.ASTNode = {
      operator: ops.get('*') as parser.Operator,
      operands: [1, {
        operator: ops.get('+') as parser.Operator,
        operands: [2, 3],
      }],
    };
    expect(shuntingYard(lexer('1 * (2 + 3)'))).toEqual(expected);
    expected = {
      operator: ops.get('*') as parser.Operator,
      operands: [1, {
        operator: ops.get('*') as parser.Operator,
        operands: [2, {
          operator: ops.get('*') as parser.Operator,
          operands: [2, {
            operator: ops.get('+') as parser.Operator,
            operands: [2, 3],
          }],
        }],
      }],
    };
    expect(shuntingYard(lexer('1 * (2 * (2 * (2 + 3)))'))).toEqual(expected);
    expect(() => {
      shuntingYard(lexer('1+ 2 * ( 3+ 1'))
    }).toThrow();
  });
  test('correctly shunts expressions with function calls', () => {
    let expected: parser.ASTNode = {
      operator: functions.get('sum') as parser.Operator,
      operands: [1, 2],
    };
    expect(shuntingYard(lexer('sum(1, 2)'))).toEqual(expected);
    expected = {
      operator: functions.get('sum') as parser.Operator,
      operands: [{
        operator: functions.get('sum') as parser.Operator,
        operands: [1, 2],
      }, 2],
    };
    expect(shuntingYard(lexer('sum(sum(1, 2), 2)'))).toEqual(expected);

    expected = {
      operator: ops.get('+') as parser.Operator,
      operands: [1, {
        operator: functions.get('sum') as parser.Operator,
        operands: [{
          operator: ops.get('+') as parser.Operator,
          operands: [1, 2],
        }, {
          operator: ops.get('*') as parser.Operator,
          operands: [1, {
            operator: ops.get('+') as parser.Operator,
            operands: [{
              operator: functions.get('sum') as parser.Operator,
              operands: [1, 2],
            }, 3],
          }],
        }],
      }],
    };
    expect(shuntingYard(lexer('1 + sum(1 + 2, 1 * (sum(1, 2) + 3))'))).toEqual(expected);
  });
});
describe('parseEquation', () => {
  /*
   test('count variables works properly', () => {
    expect(countVariables(shuntingYard(lexer('a + b')) as parser.ASTNode)).toEqual(['a', 'b']);

    expect(countVariables(shuntingYard(lexer('a + b * (1 + 2)^c')) as parser.ASTNode)).toEqual(
      ['a', 'b', 'c']
    );
  });
   */
  test('correctly parses ExplicitRectangulars and evaluates them', () => {
    let eq1 = (parser.parseEquation('z=x*y') as ExplicitRectangularEquation);
    let eq2 = (parser.parseEquation('x=y*z+1') as ExplicitRectangularEquation);
    let eq3 = (parser.parseEquation('y=(x*z+1)^5 - (7 + x)') as ExplicitRectangularEquation)
    let eq4 = (parser.parseEquation('sum(x, y*x/2 + 1)=z') as ExplicitRectangularEquation)
    for (let i = 0; i < 500; i = i + 1) {
      const mt = MersenneTwister19937.seed(1);
      const x = real(0, 10000, true)(mt);
      const y = real(0, 10000, true)(mt);
      expect(eq1.evaluate(x,y)).toEqual(x*y);
      expect(eq2.evaluate(x,y)).toEqual(x*y+1);
      expect(eq3.evaluate(x,y)).toEqual(Math.pow((x*y+1), 5) - (7+x));
      expect(eq4.evaluate(x,y)).toEqual(x+y*x/2+1);
    }
  });
  test('gives correct error for malformed ExplicitRectangulars', () => {
    // undefined var
    expect(() => (parser.parseEquation('z=x*y+a') as ExplicitRectangularEquation).evaluate(2,5)).toThrowErrorMatchingSnapshot();
    // insufficient operands
    expect(parser.parseEquation('z=x*y+') as ExplicitRectangularEquation).toEqual(new Error('insuffiecnt operands'));
    expect(parser.parseEquation('z=sum(x*y)') as ExplicitRectangularEquation).toEqual(new Error('insuffiecnt operands'));
    // not an equation
    expect(parser.parseEquation('x*y') as ExplicitRectangularEquation).toEqual(new Error('Equation does not have two expressions set equal'));
  });
});
