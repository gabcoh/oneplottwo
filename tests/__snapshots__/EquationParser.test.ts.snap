// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseEquation gives correct error for malformed ExplicitRectangulars 1`] = `"variable a is undefined"`;
